engine/internal/processor/processor.go

```diff
@@ -174,32 +159,48 @@ func (p *P) NeedLeaderElection() bool {
 	return true
 }
 
-func (p *P) run(ctx context.Context) error {
+// Start runs the processor.
+//
+// TODO(kenji): Gracefully handle an error from the server.
+func (p *P) Start(ctx context.Context) error {
+	log := p.logger.WithValues("engineID", p.engineID)
+	log.Info("Starting processor")
+	ctx = ctrl.LoggerInto(ctx, log)
 	ctx = auth.AppendWorkerAuthorization(ctx)
-	stream, err := p.client.ProcessTasks(ctx)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		_ = stream.CloseSend()
-	}()
-
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
 
-	errCh := make(chan error, 2)
-	go func() {
-		if err := p.sendEngineStatusPeriodically(ctx, stream); err != nil {
-			errCh <- fmt.Errorf("send engine status: %s", err)
+	run := func() error {
+		streamCtx, streamCancel := context.WithCancel(context.Background())
+		streamCtx = ctrl.LoggerInto(streamCtx, ctrl.LoggerFrom(ctx))
+		streamCtx = auth.AppendWorkerAuthorization(streamCtx)
+		defer streamCancel()
+		// Use separate context for the stream to gracefully handle the task requests.
+		stream, err := p.client.ProcessTasks(streamCtx)
+		if err != nil {
+			return err
 		}
-	}()
+		defer func() { _ = stream.CloseSend() }()
 
-	go func() {
-		if err := p.processTasks(ctx, stream); err != nil {
-			errCh <- fmt.Errorf("process tasks: %s", err)
+		eg, ctx := errgroup.WithContext(ctx)
+		eg.Go(func() error { return p.sendEngineStatusPeriodically(ctx, stream) })
+		eg.Go(func() error { return p.processTasks(ctx, stream) })
+		return eg.Wait()
```

Comment:
Supposet that sendEngineStatusPeriodically() returns an error. `eg.Wait` does not return the error until `processTasks` completes IIUC. Does this work when `inference-manager-server` becomes unavailable for a short period of time?

Reply:
Yes, in that case, `stream.Recv()` fails, and `processTasks` returns an error.